test/catch2_test_warp_merge_sort.cu

```diff
@@ -0,0 +1,593 @@
+    <<<1, LOGICAL_WARP_THREADS * TOTAL_WARPS>>>(thrust::raw_pointer_cast(in.data()),
+                                                thrust::raw_pointer_cast(out.data()),
+                                                segment_sizes,
+                                                oob_default,
+                                                action);
+
+  REQUIRE(cudaSuccess == cudaPeekAtLastError());
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+}
+
+/**
+ * @brief Dispatch helper function for sorting key-value pairs
+ */
+template <int ITEMS_PER_THREAD,
+          int LOGICAL_WARP_THREADS,
+          int TOTAL_WARPS,
+          typename KeyT,
+          typename ValueT,
+          typename SegmentSizesItT,
+          typename ActionT>
+void warp_merge_sort(thrust::device_vector<KeyT> &keys_in,
+                     thrust::device_vector<KeyT> &keys_out,
+                     thrust::device_vector<ValueT> &values_in,
+                     thrust::device_vector<ValueT> &values_out,
+                     SegmentSizesItT segment_sizes,
+                     KeyT oob_default,
+                     ActionT action)
+{
+  warp_merge_sort_kernel<ITEMS_PER_THREAD, LOGICAL_WARP_THREADS, TOTAL_WARPS>
+    <<<1, LOGICAL_WARP_THREADS * TOTAL_WARPS>>>(thrust::raw_pointer_cast(keys_in.data()),
+                                                thrust::raw_pointer_cast(keys_out.data()),
+                                                thrust::raw_pointer_cast(values_in.data()),
+                                                thrust::raw_pointer_cast(values_out.data()),
+                                                segment_sizes,
+                                                oob_default,
+                                                action);
+
+  REQUIRE(cudaSuccess == cudaPeekAtLastError());
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+}
+
+/**
+ * @brief Performs a stable sort on per-warp segments of data and assigns oob_default to items that
+ * are out-of-bounds.
+ */
+template <typename RandomItT, typename SegmentSizeItT, typename T>
+void compute_host_reference(RandomItT h_data,
+                            SegmentSizeItT segment_sizes,
+                            unsigned int num_segments,
+                            T oob_default,
+                            int logical_warp_items)
+{
+  for (unsigned int segment_id = 0; segment_id < num_segments; segment_id++)
+  {
+    unsigned int segment_size = segment_sizes[segment_id];
+    std::stable_sort(h_data, h_data + segment_size);
+    std::fill(h_data + segment_size, h_data + logical_warp_items, oob_default);
+    h_data += logical_warp_items;
+  }
+}
+
+/**
+ * @brief Stability requirement of the sorting algorithm
+ */
+enum class stability
+{
+  stable,
+  unstable
+};
+
+// List of key types to test
+using custom_t  = c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>;
+using key_types = c2h::type_list<std::uint8_t, std::int32_t, std::int64_t, custom_t>;
```

Comment:
Do you think there might be a value in testing custom types support only for a subset of parameters? For instance, one warp size / items per thread?